[WTF] Start removing GCC_COMPATIBLE

Constellation opened the pull request:
#### 7c2b66cb6a43f86add1fc1d5abd3301d1ae789dd
<pre>
[WTF] Start removing GCC_COMPATIBLE
<a href="https://bugs.webkit.org/show_bug.cgi?id=274364">https://bugs.webkit.org/show_bug.cgi?id=274364</a>
<a href="https://rdar.apple.com/128345506">rdar://128345506</a>

Reviewed by Ross Kirsling.

Now all supported compilers are GCC-compatible.

* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/Assertions.h:
* Source/WTF/wtf/BitSet.h:
(WTF::WordType&gt;::forEachSetBit const):
* Source/WTF/wtf/CheckedArithmetic.h:
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/MathExtras.h:
(WTF::clz):
(WTF::ctz):
(WTF::reverseBits32):
* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/PlatformCallingConventions.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WTF/wtf/StackPointer.cpp:
(WTF::currentStackPointer):
* Source/WTF/wtf/StackPointer.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::findBitInWord):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::equal):
(WTF::copyElements):

Canonical link: <a href="https://commits.webkit.org/278988@main">https://commits.webkit.org/278988@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6aa5fe6dee716d46d9a4465d6375d554fadbabc0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52148 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4568 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55421 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2862 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54452 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37876 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2569 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42436 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1834 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54244 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29115 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45004 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23507 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2279 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1030 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45496 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48288 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2424 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57018 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51656 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27273 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2478 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49834 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28507 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45116 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49068 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29418 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/63963 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7633 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28251 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12108 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -270,7 +270,7 @@
 
-#if !defined(LIKELY) && COMPILER(GCC_COMPATIBLE)
+#if !defined(LIKELY)
 #define LIKELY(x) __builtin_expect(!!(x), 1)
```
Don't we need to remove the `#if !defined(LIKELY)` block below (and likewise for the various defines that follow)?

Constellation replied:
We are keeping it since we may want to inject `LIKELY` definition outside of this file.

kmiller68 replied:
I think Ross is talking about the definition right below here:

```
#if !defined(LIKELY)
#define LIKELY(x) (x)
#endif
```

rkirsling replied:
That is indeed what I meant; there appears to be lots of unreachable code created by this patch.

rkirsling commented on Source/WTF/wtf/PrintStream.cpp:
```diff
@@ -146,7 +146,7 @@ void printInternal(PrintStream& out, unsigned char value)
 void printInternal(PrintStream& out, char16_t value)
 {
-    out.printf("%lc", value);
+    out.printf("%lc", static_cast<wint_t>(value));
```
Where'd this one come from?

Constellation replied:
With printf format warning option, this becomes warning.

rkirsling reviewed (approved):
r=me with comments

Constellation committed:
[WTF] Start removing GCC_COMPATIBLE
https://bugs.webkit.org/show_bug.cgi?id=274364
rdar://128345506

Reviewed by Ross Kirsling.

Now all supported compilers are GCC-compatible.

* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/Assertions.h:
* Source/WTF/wtf/BitSet.h:
(WTF::WordType>::forEachSetBit const):
* Source/WTF/wtf/CheckedArithmetic.h:
* Source/WTF/wtf/Compiler.h:
* Source/WTF/wtf/FastMalloc.h:
* Source/WTF/wtf/MathExtras.h:
(WTF::clz):
(WTF::ctz):
(WTF::reverseBits32):
* Source/WTF/wtf/PlatformCPU.h:
* Source/WTF/wtf/PlatformCallingConventions.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WTF/wtf/StackPointer.cpp:
(WTF::currentStackPointer):
* Source/WTF/wtf/StackPointer.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::findBitInWord):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::equal):
(WTF::copyElements):

Canonical link: https://commits.webkit.org/278988@main
